### PR TITLE
Remove arrow dependency

### LINKED ISSUE #15193: [FEA] Reduce arrow library dependencies in cudf
**Is your feature request related to a problem? Please describe.**
Arrow is in general a difficult dependency to work with, increasing build system complexity and fragility on its own while simultaneously expanding the full dependency tree, which particularly complicates use cases like conda where it leads to meaningful constraints on core system packages like protobuf, abseil, or the AWS SDK. This often hinders developer velocity when builds or CI are broken, but can also have far-reaching impacts when it creates problems with installation or running in specific environments. To prevent this, we would like to reduce or remove our dependence on Arrow libraries entirely.

Currently cudf makes use of Arrow in various ways at different levels of the stack. The primary uses of Arrow boil down to interop with host Arrow data and I/O with specific types of files. This involves interaction at both the Python layer via pyarrow, at the Cython layer (also via pyarrow), and in C++. Both Cython and C++ interactions are particularly problematic because they involve C-level interactions, which sets ABI-level constraints that are significantly tighter than we would like while also significantly complicating build (CMake, Python builds) and packaging (narrow Arrow version support ranges leading to limited support of other packages in the dependency tree). Python interactions are generally less difficult to work around, especially since Python code can be written to dynamically adapt to the pyarrow version.

**Describe the solution you'd like**

We should look to remove the Arrow dependencies from the various layers of cudf (Java, Python, Cython, C++) to the greatest extent possible, ideally entirely.

For Arrow Array interop code, this can be accomplished by using the Arrow C Data Interface (see #5097), which provides an ABI-stable way to interchange Arrow data without directly using Arrow libraries. To make this even easier, the [nanoarrow](https://github.com/apache/arrow-nanoarrow) library was created to support clients that wish to produce or interpret [Arrow C Data](https://arrow.apache.org/docs/format/CDataInterface.html) and [Arrow C Streams](https://arrow.apache.org/docs/format/CStreamInterface.html) structures, without having to include a dependency on libarrow. We can make use of that (see also #13678 which discusses this in depth). For Python interaction we can use [Arrow's pycapsule interface](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html), which provides a standard way to interchange this data from Python. We can write Cython code leveraging this interface to get Arrow C Data from pyarrow objects without relying directly on pyarrow's Cython, therefore also allowing us to remove this dependency from the Cython layer.

For I/O, the question is a bit trickier. We currently have limited usage of libarrow headers in our C++, and those features largely exist only for Python support for reading Arrow's NativeFiles. We could in principle remove those from the C++ entirely, which would in turn allow us to remove libarrow as a dependency of libcudf. However, libcudf tests would still need libarrow (removing that dependency would require significant additional work). Moreover, those features would still be used by cudf Cython, so we would just be limiting the dependency. However, this could at least allow us to remove Arrow as a build-time dependency for both libcudf and the low-level pylibcudf Python API (#13921) that we are currently developing, which would still be a significant improvement since it would avoid imposing the Arrow dependency on low-level consumers of our APIs at the Python level. Then we could come back to working on replacing the cudf Cython usage.

Based on the above, the current plan is the following:
1. Remove libarrow as a dependency of libcudf/pylibcudf:
    a. Remove the compiled parts of `arrow_io_source.cpp` and make `arrow_io_source.hpp` a standalone header not compiled by anything in libcudf.
    b. Rewrite cudf Cython to use the arrow headers directly.
    c. Add new interop code that uses the Arrow C Data interface (see #15047)
    d. Rewrite Python interop code to call through to the new interfaces
    e. Remove the old Cython bindings for interop
2. Remove pyarrow Cython linkages from cudf Cython
    a. This will require some exploration as to how we can maintain performant file reading. We may have to implement our own minimal version of something like Arrow's NativeFile reader interface.
    b. Once the above is done, we'll need to rewrite cuIO C++ to consume this interface and remove the current functions.
3. Rewrite libcudf tests to remove libarrow dependence.
    a. This will require further investigation into how tests could be rewritten without Arrow. One possibility would be rewriting these tests as pylibcudf tests (see #15133) that use pyarrow instead (only the Python API, no Cython). That would give us access to the same functionality without tying us to linking to the libarrow library

**Additional context**

Code pointers where libarrow is used in 24.04
| Source file | Arrow include | Notes |
|---|---|---|
| `detail/interop.hpp` | `api.h` | `to_arrow_array` uses many array classes: `arrow::*Array`, `arrow::TimeUnit::*`, `arrow::*Type` also `arrow::MemoryPool`, `arrow::Scalar`, `arrow::Table`. I believe all of these are covered by nanoarrow |
| `include/cudf/interop.hpp`  | `api.h` | uses `arrow::Table`, `arrow::MemoryPool`, `arrow::default_memory_pool`, `arrow::Scalar`. I believe all of these are covered by nanoarrow |
| `include/cudf/io/arrow_io_source.hpp` | `filesystem/filesystem.h` <br> `io/interfaces.h`  | uses `arrow::io:RandomAccessFile`, `arrow::fs::FileSystem`. See #13698 for the work to refactor `arrow_io_source` out of `datasource` |
| `include/cudf/io/arrow_io_source.cpp` | `buffer.h` <br> `filesystem/filesystem.h` <br> `result.h`  | uses `arrow::Buffer`, `arrow::fs::FileSystemFromUri`, |
| `src/io/utilities/datasource.cpp` | `io/memory.h` | to be solved by #15189 |

| Test file | Arrow include | Notes |
|---|---|---|
| `tests/interop/arrow_utils.hpp` | `util/bitmap_builders.h` for `arrow::internal::BytesToBits` | Also uses many arrow types such as: `arrow::Array`, `arrow:DictionaryArray`, `arrow::dictionary`, `arrow::Table`,  `arrow::Decimal128Builder`, `arrow::decimal`, `arrow::default_memory_pool`, `arrow::ListArray`, `arrow::list` , `arrow::Buffer`, `arrow::StringBuilder`, `arrow::StringArray` , `arrow::BooleanArray`, `arrow::BooleanBuilder` <br> needs research - can all of these references be migrated to nanoarrow? |
| `tests/io/arrow_io_source_test.cpp`  | `io/api.h`  <br> `filesystem/filesystem.h` <br> `filesystem/s3fs.h` <br> `util/config.h` | uses `arrow::fs::FileSystemFromUri`, `arrow::fs::EnsureS3Finalized` |
| `tests/io/json_test.cpp` | `io/api.h` | Uses `arrow::io::ReadableFile` as part of a test for reading from an `ArrowFileSource` |
| `tests/io/csv_test.cpp` | `io/api.h` | uses `arrow::io::ReadableFile` |
| `tests/quantiles/percentile_approx_test.cpp` | `util/tdigest.h` | uses `arrow::internal::TDigest`. presumably we could replace this with our own limited implementation |

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR removes libarrow as a dependency of libcudf since we no longer use any of its APIs in our C++ code. The following places remain dependent on libarrow:
- tests: We have tests demonstrating how to interoperate with libarrow objects, as well as other tests that leverage Arrow for I/O.
- examples: We have an example demonstrating interop with libarrow arrays.
- JNI: The JNI is still using libarrow to handle ingestion or production of Arrow buffers.

In all three cases above, we are now statically linking libarrow. We also always pull it in via CPM, which means that we never require libarrow to exist on the user's system anymore. Of the above three cases, we should expect the first two to persist indefinitely. The JNI could be updated to use nanoarrow instead if desired, but that is not critical since the primary benefit of removing libarrow as a direct dependency is to remove it as a constraint for package managers such as conda in environments where we must match the version of Arrow required by other dependencies.

pyarrow remains a dependency of the cudf Python packages. For now, this PR retains the tight pinning on 16.1 since we know that this version works. A future PR will loosen this pinning since we are no longer constrained to ABI-compatible versions and can support a range of pyarrow versions that support the necessary Python APIs (I believe pyarrow>=13 will work, but that remains to be tested).

Resolves #15193 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]
Looks good to me.

@vyasr Can we order this after https://github.com/rapidsai/cudf/pull/15483? I think the conflicts will be hard to understand otherwise and I don't want this change to delay #15483.

**Review by jameslamb — APPROVED** [review]
🎉

File: cpp/cmake/thirdparty/get_arrow.cmake [hunk 1]
```diff
@@ -174,23 +80,21 @@ function(find_and_configure_arrow VERSION BUILD_STATIC ENABLE_S3 ENABLE_ORC ENAB
...
     GIT_TAG apache-arrow-${VERSION}
     GIT_SHALLOW TRUE SOURCE_SUBDIR cpp
+    EXCLUDE_FROM_ALL ${EXCLUDE_FROM_ALL}
     OPTIONS "CMAKE_VERBOSE_MAKEFILE ON"
             "ARROW_ACERO ON"
             "ARROW_IPC ON"
             "ARROW_DATASET ON"
             "ARROW_WITH_BACKTRACE ON"
             "ARROW_CXXFLAGS -w"
             "ARROW_JEMALLOC OFF"
```
[the remark this answers is not in the document]
Comment (vyasr) [reply cpp/cmake/thirdparty/get_arrow.cmake:90]: Yes that is now on line 191 in this file: https://github.com/rapidsai/cudf/pull/16640/files#diff-37169a2940bf7bc08ba411474ff544309f1baadbb3c0c3713efbb6d82c0b594fR191

[the remark this answers is not in the document]
Comment (robertmaynard) [reply cpp/cmake/thirdparty/get_arrow.cmake:90]: Thanks! That was hard to see in the diff view

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

File: cpp/cmake/thirdparty/get_arrow.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply cpp/cmake/thirdparty/get_arrow.cmake:90]: Yup no worries, I can tell.

File: conda/recipes/cudf/meta.yaml [hunk 2]
```diff
@@ -83,8 +81,8 @@ requirements:
     - pandas >=2.0,<2.2.3dev0
     - cupy >=12.0.0
     - numba >=0.57
-    - numpy >=1.23,<3.0a0
-    - {{ pin_compatible('pyarrow', max_pin='x.x') }}
+    - numpy >=1.23,<2.0a0
```
Comment (jakirkham) [inline conda/recipes/cudf/meta.yaml:84]: This should still be NumPy 3 as the upper bound

```suggestion
    - numpy >=1.23,<3.0a0
```

File: conda/recipes/pylibcudf/meta.yaml [hunk 3]
```diff
@@ -80,8 +78,8 @@ requirements:
     - python
     - typing_extensions >=4.0.0
     - pandas >=2.0,<2.2.3dev0
-    - numpy >=1.23,<3.0a0
-    - {{ pin_compatible('pyarrow', max_pin='x.x') }}
+    - numpy >=1.23,<2.0a0
```
Comment (jakirkham) [inline conda/recipes/pylibcudf/meta.yaml:81]:
```suggestion
    - numpy >=1.23,<3.0a0
```

**Review by jakirkham — CHANGES_REQUESTED** [review]
Have noted a few changes that were missed below

**Comment by vyasr** [comment]
Fixed numpy pinnings.

**Review by jakirkham — APPROVED** [review]
Looks good. Thanks Vyas! 🙏

**Comment by vyasr** [comment]
Waiting on some final confirmation that https://github.com/NVIDIA/spark-rapids-jni/pull/2357 works for the plugin and then this should be good to go.

File: java/src/main/native/CMakeLists.txt [hunk 4]
```diff
@@ -212,6 +212,10 @@ target_compile_definitions(
 )
 target_link_options(cudfjni PRIVATE "-Wl,--no-undefined")
 
+set(CUDF_ENABLE_ARROW_PARQUET ON)
+include(../../../../cpp/cmake/thirdparty/get_arrow.cmake)
+target_link_libraries(cudfjni PRIVATE ${ARROW_LIBRARIES})
+
```
Comment (mythrocks) [inline java/src/main/native/CMakeLists.txt:218]: 👍

**Review by mythrocks — APPROVED** [review]
Thanks for working on this, @vyasr.  Approving, on behalf of CUDF JNI.

**Comment by ttnghia** [comment]
I still see `cudf/cpp/cmake/thirdparty/get_arrow.cmake`. Is this needs to be removed too?

**Comment by vyasr** [comment]
No. We still use arrow in tests, examples, and the JNI, and that CMake code is used in all three. In all three cases we now build and statically link to Arrow as part of the build, though, so there is no longer any library dependency in those cases. The critical change in this PR is removing the dependency from libcudf and the cudf Python modules.

**Pull request merged**